Trest integration

LordMax2 opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Integration with the Trest Solar Controller.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

New integration.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://docs.trest.se

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works. 

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [X] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

LordMax2 committed:
Init of new component

LordMax2 committed:
Restructured to follow SOLID a little bit more

LordMax2 committed:
First working implementation of the integration

LordMax2 committed:
Restructured sensor

LordMax2 committed:
Fixed scan interval

joostlek commented:
Can we put the domain name to something better? 😂

autinerd commented:
Hi, thanks for your contribution! 😊 As @joostlek already said, the integration name is not really good. Can you please rename it to something like `trest` or `trest_solar`, before we can start beginning with the review?

LordMax2 committed:
Update const.py

LordMax2 committed:
Update manifest.json

LordMax2 committed:
Renamed directory

LordMax2 committed:
Renamed models directory

autinerd commented:
> Hi! So I usually do domain driven design on my applications, but if you prefer trest_solar instead I wouldn't mind changing it. :)

I understand, but because the code will become part of Home Assistant, the `homeassistant` in the name is redundant :D You still need to rename the test directory.

LordMax2 commented:
> > Hi! So I usually do domain driven design on my applications, but if you prefer trest_solar instead I wouldn't mind changing it. :)
> 
> I understand, but because the code will become part of Home Assistant, the `homeassistant` in the name is redundant :D You still need to rename the test directory.

Sorry, I realized that the domain driven design goes under the HomeAssistant and not my application, I cannot just use my domain design in a conpletley different applications, that would just be mayhem and chaotic if everyone had their own domain in the HomeAssistant component 😂 I was just me who was slow to realize 🐌 my bad :) 

I will change the test directory right away!

LordMax2 committed:
Renamed tests directory and domain in test files

zweckj reviewed (changes requested):
thanks for your contribution! First feedback for you

LordMax2 commented on homeassistant/components/trest_solar/services/trest_identity_service.py:
```diff
@@ -0,0 +1,65 @@
+import base64  # noqa: D100
```
What do you mean by this?

zweckj replied:
a right, just noticed you don't have a library yet. We require the actual device logic (REST calls, etc.) to be in a separate PyPi package outside of Home Assistant. Home Assistant should only use that library, but not contain such connection implementation. This lib is then included in your manifest's `requirements`

LordMax2 commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,72 @@
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate if the user input contains valid credentials."""
+
+    identity = TrestIdentityService(hass, data[CONF_USERNAME], data[CONF_PASSWORD])
+    await identity.authenticate_async()
+
+    if not identity.check_token_is_expired():
+        raise InvalidAuth
+
+    return {"title": "Trest Solar Controller"}
```
Do you mean move the "check" into authenticate?

zweckj replied:
if you just move the whole function inline, you'd have things inside the `try` that can't fail (e.g.) the identityservice initialization. The goal is to only have code inside the try which can raise errors.

LordMax2 commented on homeassistant/components/trest_solar/coordinator.py:
```diff
@@ -0,0 +1,63 @@
+        """Fetch data from Trest Cloud."""
+
+        _LOGGER.info("Requesting Solar History data from Cloud Trest")
+        response = await self.trest_solar_service.get_latest_solar_history_async()
```
Yes, it can, how would I handle this best? If the user is unauthorized, timeout or the service is down, then it would throw an exception, do you have any tips?

zweckj replied:
You would raise `ConfigEntryAuthFailed` on auth errors to trigger a reauthentication flow (but add that in a separate PR later, for now you'd raise `ConfigEntryError`) and `UpdateFailed` on generic update errors (`from homeassistant.helpers.update_coordinator import UpdateFailed`)

LordMax2 commented on homeassistant/components/trest_solar/coordinator.py:
```diff
@@ -0,0 +1,63 @@
+
+        return solar_history
+
+    async def async_config_entry_first_refresh(self) -> None:
+        """Refresh data for the first time when a config entry is setup.
+
+        Will automatically raise ConfigEntryNotReady if the refresh
+        fails. Additionally logging is handled by config entry setup
+        to ensure that multiple retries do not cause log spam.
+        """
+        await self._async_refresh(
+            log_failures=False, raise_on_auth_failed=True, raise_on_entry_error=True
+        )
+        if self.last_update_success:
+            return
+        ex = ConfigEntryNotReady()
+        ex.__cause__ = self.last_exception
+        raise ex
+
+    async def async_refresh(self) -> None:
+        """Refresh data and log errors."""
+        await self._async_refresh(log_failures=True)
```
Ah, missed that

LordMax2 commented on homeassistant/components/trest_solar/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "codeowners": ["@LordMax2"],
+  "config_flow": true,
+  "dependencies": [],
+  "documentation": "https://docs.trest.se/home-automation/home-assistant",
```
Ok 👍

LordMax2 commented on homeassistant/components/trest_solar/models/solar_history.py:
```diff
@@ -0,0 +1,46 @@
+"""Class file for the entity SolarHistory."""
```
What do you mean by library?

LordMax2 commented on homeassistant/components/trest_solar/services/trest_solar_service.py:
```diff
@@ -0,0 +1,31 @@
+import json  # noqa: D100
```
What do you mean with library? :)

LordMax2 committed:
Reviewed review, some more follow up questions

LordMax2 committed:
Fixed fetching credentials in Coordinator init for CloudSolarTrestService

LordMax2 commented:
> thanks for your contribution! First feedback for you

Thank you for your review and patience! I have fixed some of the issues and asked some follow up questions :)

zweckj commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,72 @@
+    identity = TrestIdentityService(hass, data[CONF_USERNAME], data[CONF_PASSWORD])
+    await identity.authenticate_async()
+
+    if not identity.check_token_is_expired():
```
Just to be sure: This returns `False` when the token is expired? (Judging by the name I would assume it returns `True` when the token is expired)

LordMax2 committed:
Moved services to pypi packages

LordMax2 committed:
Fixed token inversion

LordMax2 commented:
I have now moved logic that is supposed to be in libraries!

emontnemery committed:
Merge branch 'dev' into trest-integration

emontnemery reviewed (changes requested):
Please address the comments from previous reviews. I also added some additional comments.

LordMax2 commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,110 @@
+) -> None:
+    """Add a Trest Solar Controller entry."""
+    coordinator: TrestDataCoordinator = hass.data[DOMAIN][entry.entry_id]
+    descriptions = [
```
What do you mean with "file level"?

joostlek replied:
Move it out of the function and declare it as a list

LordMax2 replied:
I am confused, what am I supposed to move out here?

joostlek replied:
The list of entity descriptions, check airgradient

LordMax2 commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,70 @@
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+            except CannotConnect:
```
Ok, I think it was some kind of default when I added the config flow, removed it now!

LordMax2 commented on homeassistant/components/trest_solar/coordinator.py:
```diff
@@ -0,0 +1,48 @@
+            update_interval=timedelta(seconds=REFRESH_INTERVAL),
+        )
+
+        if self.config_entry is not None:
```
The ruff warns because the configentry has the type "ConfigEntry | None" in this context

LordMax2 commented on homeassistant/components/trest_solar/__init__.py:
```diff
@@ -0,0 +1,35 @@
+    coordinator = TrestDataCoordinator(hass, _LOGGER)
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```
I did not manage to get this working, ruff is complaining

joostlek replied:
Please push what you have

LordMax2 commented on homeassistant/components/trest_solar/const.py:
```diff
@@ -0,0 +1,4 @@
+"""Constants for the TrestSolarController integration."""
+
+DOMAIN = "trest_solar"
+REFRESH_INTERVAL = 30
```
Fixed!

LordMax2 commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,110 @@
+    descriptions = [
+        SensorEntityDescription(
+            key="battery_discharge",
+            name="Battery Discharge",
```
How do I implement strings.json? I can only find guides about the YAML flows..

LordMax2 commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,70 @@
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                return self.async_create_entry(title=info["title"], data=user_input)
```
How can I check if there are any added user with the same username already? :)

joostlek replied:
Is there something truly unique we know about the user? Preferably an unique user identifier. Otherwise we'd have to use the username, and then we have to know if the username is changeable, if so, we have to do another approach

LordMax2 replied:
The unique thing is the username, it is the email of the user, and that is always unique

joostlek replied:
Is it changeable?

LordMax2 replied:
Nope

joostlek replied:
Then set that as the unique id of the config flow

LordMax2 replied:
How do I set this to unique?

LordMax2 committed:
Fixed some of the comments

LordMax2 committed:
Merge branch 'trest-integration' of https://github.com/LordMax2/TrestSolarController-HomeAssistant2 into trest-integration

LordMax2 commented:
> You have not pushed since the last feedback, is that correct?

Sorry, I just realized, I thought I pushed weeks ago, pushed it now!

joostlek commented on homeassistant/components/trest_solar/coordinator.py:
```diff
@@ -0,0 +1,48 @@
+class TrestDataCoordinator(DataUpdateCoordinator[SolarHistory]):
+    """Trest coordinator."""
+
+    def __init__(self, hass: HomeAssistant, logger: logging.Logger) -> None:
```
Why is the logger a parameter?

joostlek commented on homeassistant/components/trest_solar/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "codeowners": ["@LordMax2"],
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/trest_solar",
+  "homekit": {},
```
Remove empty fields

LordMax2 replied:
Fixed!

joostlek commented on homeassistant/components/trest_solar/manifest.json:
```diff
@@ -0,0 +1,10 @@
+  "documentation": "https://www.home-assistant.io/integrations/trest_solar",
+  "homekit": {},
+  "iot_class": "cloud_polling",
+  "requirements": ["trest_identity==1.0.0", "trest_solar==1.0.1"]
```
Do they need loggers?

LordMax2 replied:
They do not need loggers, I used two libs to create one for each layer of application, but of course the identity should be included in the solar one, because otherwise the solar one isnt going to work.

LordMax2 commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,71 @@
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
+    """Validate if the user input contains valid credentials."""
+
+    identity = TrestIdentityService(hass, data[CONF_USERNAME], data[CONF_PASSWORD])
+    try:
+        await identity.authenticate_async()
```
Im not sure what you mean by this? Skip the try catch?

joostlek replied:
No, merge this and async step user

LordMax2 commented on homeassistant/components/trest_solar/coordinator.py:
```diff
@@ -0,0 +1,48 @@
+from __future__ import annotations  # noqa: D100
```
Not sure why this import is here, removed now!

LordMax2 commented on homeassistant/components/trest_solar/coordinator.py:
```diff
@@ -0,0 +1,48 @@
+            update_interval=timedelta(seconds=REFRESH_INTERVAL),
+        )
+
+        if self.config_entry is not None:
```
Where?

LordMax2 commented:
I get this error when I try the custom ConfigEntry:

`homeassistant/components/trest_solar/__init__.py:13: error: invalid syntax; you likely need to run mypy using Python 3.12 or newer  [syntax]
Found 1 error in 1 file (errors prevented further checking)`

How can I upgrade the python version of the dev container? I use python 3.12 on my computer regulary

LordMax2 committed:
Removed unnecesary logger
Changed ConfigEntry to take a custom coordinator in runtime_data
Cleanup unnecesary imports
Upgraded trest_solar library to return the final type immediatley instead of retruning a string and converting the result

joostlek commented:
No clue, maybe rebase on dev and rebuild container?

LordMax2 committed:
Inline validate_input in step_user

LordMax2 commented:
> No clue, maybe rebase on dev and rebuild container?

Solved it via not using the "type" prefix

joostlek commented:
But we should use the type prefix :)

joostlek commented:
But like, it says you're not running mypy in 3.12, so it looks like _something_ is off

LordMax2 committed:
New version of trest_solar in requirements
Cleaned up sensor types descriptions
Fixed units of SensorEntityDescriptions

LordMax2 commented:
Now everything except the "unique id entry" thing, how do I solve that one?

LordMax2 committed:
Merge branch 'dev' into trest-integration

LordMax2 committed:
Merge branch 'dev' into trest-integration

joostlek commented:
`self.async_set_unique_id` and after that the abort. Do a global search and you'll find some

joostlek commented:
Not sure how much time you have nowadays, but I think we can definitely aim this for 2024.8

LordMax2 committed:
Update manifest.json to include the latest version of trest_solar
Modify async_step_user to abort if unique id is already configured

LordMax2 committed:
Add abort_if_uid_configured in configflow

joostlek commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,67 @@
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+        if user_input is not None:
+            try:
```
Only have stuff in the try block that can raise

joostlek commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,67 @@
+        )
+
+
+class CannotConnect(HomeAssistantError):
+    """Error to indicate we cannot connect."""
+
+
+class InvalidAuth(HomeAssistantError):
+    """Error to indicate there is invalid auth."""
```
Unused

joostlek commented on homeassistant/components/trest_solar/coordinator.py:
```diff
@@ -0,0 +1,41 @@
+
+
+class TrestDataCoordinator(DataUpdateCoordinator[SolarHistory]):
+    """Trest coordinator."""
+
```
```suggestion
    """Trest coordinator."""

    config_entry: ConfigEntry
```

joostlek replied:
```suggestion
    """Trest coordinator."""

    config_entry: ConfigEntry
```

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,133 @@
+        key="battery_discharge",
+        name="Battery Discharge",
+        device_class=SensorDeviceClass.POWER,
+        native_unit_of_measurement="kW",
```
Please use constants for this. UnitOfPower.KILO_WATT

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,133 @@
+SENSOR_TYPES: tuple[SensorEntityDescription, ...] = (
+    SensorEntityDescription(
+        key="battery_discharge",
+        name="Battery Discharge",
```
Please use translation keys

LordMax2 replied:
What is translation keys? How do I implement them?

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,133 @@
+        name="Battery Discharge",
+        device_class=SensorDeviceClass.POWER,
+        native_unit_of_measurement="kW",
+        icon="mdi:battery-arrow-down",
```
And use icon translations

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,133 @@
+    """Add a Trest Solar Controller entry."""
+    coordinator: TrestDataCoordinator = entry.runtime_data
+
+    entities = [
+        TrestSolarControllerSensor(coordinator, entry.entry_id, description)
+        for description in SENSOR_TYPES
+    ]
+
+    async_add_entities(entities)
```
Can be merged

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,133 @@
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Add a Trest Solar Controller entry."""
+    coordinator: TrestDataCoordinator = entry.runtime_data
```
```suggestion
    coordinator = entry.runtime_data
```

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,133 @@
+    def __init__(
+        self,
+        coordinator: TrestDataCoordinator,
+        entry_id: str,
```
Since the username is our unique thing, should we use the config entry unique id as parameter? It's `str|None` by default so you need to `assert entry.unique_id` first

LordMax2 replied:
Is this needed? Isnt entry_id enough or is that something different?

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,133 @@
+        )
+
+    @property
+    def native_value(self) -> str | int | float | datetime | None:
```
```suggestion
    def native_value(self) -> StateType | datetime:
```

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,133 @@
+        native_unit_of_measurement="kW",
+        icon="mdi:white-balance-sunny",
+    ),
+    SensorEntityDescription(
+        key="timestamp",
+        name="Timestamp",
+        device_class=SensorDeviceClass.TIMESTAMP,
+        icon="mdi:clock",
+    ),
```
We don't allow sensors that show the last updated time since HA already stores that data.

LordMax2 replied:
This is kind of needed because this is the timestamp of when the data was pushed from the IoT device that fetches the data and pushes it to the cloud.

LordMax2 committed:
Change units, cleanup

LordMax2 committed:
Merge branch 'dev' into trest-integration

LordMax2 commented:
Sorry for the late PR corrections, I have some questions now with the translation keys, was there something else more other than that?

joostlek commented:
https://developers.home-assistant.io/blog/2023/03/27/entity_name_translations?_highlight=translation#translating-entity-name

LordMax2 committed:
Add icon translations
Add state name translations

joostlek commented on homeassistant/components/trest_solar/strings.json:
```diff
@@ -17,5 +17,33 @@
+  "entity": {
+    "sensor": {
+      "battery_discharge": {
+        "name": "Battery Discharge"
+      },
+      "battery_charge": {
+        "name": "Battery Charge"
+      },
+      "battery_capacity": {
+        "name": "Battery Capacity"
+      },
+      "battery_stored_power": {
+        "name": "Battery Stored Power"
+      },
+      "total_load_active_power": {
+        "name": "Total Load Active Power"
+      },
+      "realtime_solar": {
+        "name": "Realtime Solar"
+      },
+      "solar_profile": {
+        "name": "Solar Profile"
+      },
+      "daily_yeild": {
+        "name": "Daily Yeild"
```
Please use `Sentence case`

LordMax2 committed:
Change translation keys to "sentence case"

joostlek commented on homeassistant/components/trest_solar/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+TrestConfigEntry = ConfigEntry[TrestDataCoordinator]
```
```suggestion
type TrestConfigEntry = ConfigEntry[TrestDataCoordinator]
```

joostlek commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+
+    async def async_step_user(
+        self,
+        user_input: Optional[dict[str, Any]] = None,  # noqa: UP007
```
please use new optional notation
```suggestion
        user_input: dict[str, Any] | None = None,
```

joostlek commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+                user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+            )
+
+            info = {"title": "Trest Solar Controller"}
```
this doesn't add much, might as well just move that string to the create_entry call

joostlek commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+            else:
+                return self.async_create_entry(title=info["title"], data=user_input)
+
+            # Ensure this only happens if user_input is not None
+            await self.async_set_unique_id(user_input[CONF_USERNAME])
+            self._abort_if_unique_id_configured()
```
We need to do this before the `async_create_entry`

LordMax2 committed:
Merge branch 'dev' into trest-integration

joostlek commented on homeassistant/components/trest_solar/config_flow.py:
```diff
@@ -0,0 +1,66 @@
+                    raise InvalidAuth
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
```
Can't it catch more?

joostlek commented on homeassistant/components/trest_solar/coordinator.py:
```diff
@@ -0,0 +1,41 @@
+            logger=_LOGGER,
+        )
+
+        if self.config_entry is not None:
+            self.trest_solar_service = CloudSolarTrestService(
+                self.config_entry.data["username"],
+                self.config_entry.data["password"],
+            )
+        else:
+            raise MissingConfigEntryException("Missing config entry")
```
```suggestion
        self.trest_solar_service = CloudSolarTrestService(
            self.config_entry.data[CONF_USERNAME],
            self.config_entry.data[CONF_PASSWORD],
        )
```

joostlek commented on homeassistant/components/trest_solar/coordinator.py:
```diff
@@ -0,0 +1,41 @@
+        return await self.trest_solar_service.get_latest_solar_history_async()
+
+
+class MissingConfigEntryException(Exception):
```
can be removed

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,122 @@
+from .const import DOMAIN
+from .coordinator import TrestDataCoordinator
+
+SENSOR_TYPES: tuple[SensorEntityDescription, ...] = (
```
Do these need stateclasses?

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,122 @@
+        key="daily_yeild",
+        translation_key="daily_yeild",
+        device_class=SensorDeviceClass.ENERGY,
+        native_unit_of_measurement="kW",
```
Please use constants for the unit

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,122 @@
+        device_class=SensorDeviceClass.POWER,
+        native_unit_of_measurement=UnitOfPower.KILO_WATT,
+    ),
+    SensorEntityDescription(
+        key="solar_profile",
+        translation_key="solar_profile",
+        device_class=SensorDeviceClass.POWER,
+    ),
```
What does this sensor do, it's missing a unit

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,122 @@
+    """Add a Trest Solar Controller entry."""
+    coordinator = entry.runtime_data
+
+    async_add_entities(
+        [
+            TrestSolarControllerSensor(coordinator, entry.entry_id, description)
+            for description in SENSOR_TYPES
+        ]
+    )
+
```
```suggestion
    async_add_entities(
        TrestSolarControllerSensor(coordinator, entry.entry_id, description)
        for description in SENSOR_TYPES
    )

```

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,122 @@
+    """The sensor for Trest Solar Controller."""
+
+    _attr_has_entity_name = True
+    entity_description: SensorEntityDescription
```
```suggestion
```
Is the default

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,122 @@
+
+        key = self.entity_description.key
+
+        if key == "timestamp":
+            timestamp_str = self.coordinator.data.to_dict().get(key)
+            if timestamp_str:
+                return datetime.fromisoformat(timestamp_str)
```
there is no timestamp entity

joostlek commented on homeassistant/components/trest_solar/sensor.py:
```diff
@@ -0,0 +1,122 @@
+            if timestamp_str:
+                return datetime.fromisoformat(timestamp_str)
+
+        return self.coordinator.data.to_dict().get(key)
```
why to dict? If the source data is typed, please extend SensorEntityDescription and add a function `value_fn`. You can then make the calls typesafe. Check `airgradient` for an example

joostlek commented on homeassistant/components/trest_solar/strings.json:
```diff
@@ -0,0 +1,49 @@
+    "step": {
+      "user": {
+        "data": {
+          "host": "[%key:common::config_flow::data::host%]",
+          "username": "[%key:common::config_flow::data::username%]",
+          "password": "[%key:common::config_flow::data::password%]"
```
Can we use `data_description` to give more context?

LordMax2 replied:
Where will this data_description be shown in the GUI? Is it when you are adding a new entry, you get some kind of description?

joostlek replied:
It's a label underneath the input field

joostlek commented on homeassistant/components/trest_solar/strings.json:
```diff
@@ -0,0 +1,49 @@
+      "unknown": "[%key:common::config_flow::error::unknown%]"
+    },
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
```suggestion
      "already_configured": "[%key:common::config_flow::abort::already_configured_account%]"
```

LordMax2 committed:
Cleanup, reviewed joostlek's review

autinerd commented on homeassistant/components/trest_solar/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    return True
```
```suggestion
    return True


async def async_unload_entry(hass: HomeAssistant, entry: TrestConfigEntry) -> bool:
    """Unload a config entry."""
    return await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
```
The unload function is missing.

autinerd commented on homeassistant/components/trest_solar/manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "name": "TrestSolarController",
+  "codeowners": ["@LordMax2"],
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/trest_solar",
+  "iot_class": "cloud_polling",
```
```suggestion
  "documentation": "https://www.home-assistant.io/integrations/trest_solar",
  "integration_type": "device",
  "iot_class": "cloud_polling",
```

LordMax2 committed:
Update homeassistant/components/trest_solar/__init__.py

Co-authored-by: Sid <27780930+autinerd@users.noreply.github.com>
